marconi-sidechain/src/Marconi/Sidechain/Error.hs

```diff
@@ -16,3 +29,48 @@ data QueryExceptions
   | IndexerInternalError !Text
   deriving stock (Show)
   deriving anyclass (Exception)
+
+-- * Mapping of exceptions to exit codes
+
+class HasExitCode e where
+  toExitCode :: e -> Int
+
+instance HasExitCode (IndexerError Void) where
+  toExitCode = indexerErrorToExit
+
+indexerErrorToExit :: IndexerError Void -> Int
+indexerErrorToExit err =
+  case err of
+    Timeout _ -> 124
+    _ -> 1
+
+-- * Signal handling
+
+{- | Takes an action and runs it while waiting for either a `SIGINT` or a `SIGTERM` to be thrown.
+	If one is thrown, it's swallowed, but the value is returned to be interpreted by the caller of
+	this function.
+-}
+withSignalHandling :: IO a -> IO (Either Signal a)
+withSignalHandling action = do
+  termVar <- newEmptyMVar
+  let terminate terminationType = void $ tryPutMVar termVar terminationType
+      waitForTermination = takeMVar termVar
+      signals = [sigINT, sigTERM]
+
+  traverse_ (\signal -> installHandler signal (CatchOnce (terminate signal)) Nothing) signals
+  race
+    waitForTermination
+    action
+
+{- | Maps a signal to an exit code.
+
+		Maps SIGINT (2) to 130
+		Maps SIGTERM (15) to 143
+		Maps all other signals to 137 (SIGKILL)
+-}
+signalToExit :: Int -> Int
+signalToExit signal =
+  case signal of
+    2 -> 130
+    15 -> 143
+    _ -> 137
```

Comment:
Why didn't you take a `Signal` as an input, together with guards, it could lead to the following

```
signalToExit :: Signal -> Int
signatToExit signal
  | signal == sigINT = 130
  | signal == sigTERM = 143
  | otherwise = 137
```

(we might event introduce constant to provide semantic for `130`, `143` and `137`)